ecosystem/typescript/sdk/src/transactions/account_sequence_number.ts

```diff
@@ -0,0 +1,172 @@
+ * If ever waiting more than `maxWaitTime` restart the sequence number to the current on-chain state
+ *
+ * Assumptions:
+ * Accounts are expected to be managed by a single AccountSequenceNumber and not used otherwise.
+ * They are initialized to the current on-chain state, so if there are already transactions in
+ * flight, they may take some time to reset.
+ * Accounts are automatically initialized if not explicitly
+ *
+ * Notes:
+ * This is co-routine safe, that is many async tasks can be reading from this concurrently.
+ * The state of an account cannot be used across multiple AccountSequenceNumber services.
+ * The synchronize method will create a barrier that prevents additional nextSequenceNumber
+ * calls until it is complete.
+ * This only manages the distribution of sequence numbers it does not help handle transaction
+ * failures.
+ * If a transaction fails, you should call synchronize and wait for timeouts.
+ */
+
+import { AptosAccount } from "../account";
+import { Provider } from "../providers";
+import { sleep } from "../utils";
+
+// returns `now` time in seconds
+const now = () => Math.floor(Date.now() / 1000);
+
+export class AccountSequenceNumber {
+  readonly provider: Provider;
+
+  readonly account: AptosAccount;
+
+  // sequence number on chain
+  lastUncommintedNumber: bigint | null = null;
+
+  // local sequence number
+  currentNumber: bigint | null = null;
+
+  /**
+   * We want to guarantee that we preserve ordering of workers to requests.
+   *
+   * `lock` is used to try to prevent multiple coroutines from accessing a shared resource at the same time,
+   * which can result in race conditions and data inconsistency.
+   * This code actually doesn't do it though, since we aren't giving out a slot, it is still somewhat a race condition.
+   *
+   * The ideal solution is likely that each thread grabs the next number from a incremental integer.
+   * When they complete, they increment that number and that entity is able to enter the `lock`.
+   * That would guarantee ordering.
+   */
+  lock = false;
+
+  maxWaitTime: number;
+
+  maximumInFlight: number;
+
+  sleepTime: number;
+
+  constructor(
+    provider: Provider,
+    account: AptosAccount,
+    maxWaitTime: number,
+    maximumInFlight: number,
+    sleepTime: number,
+  ) {
+    this.provider = provider;
+    this.account = account;
+    this.maxWaitTime = maxWaitTime;
+    this.maximumInFlight = maximumInFlight;
+    this.sleepTime = sleepTime;
+  }
+
+  /**
+   * Returns the next available sequence number for this account
+   *
+   * @returns next available sequence number
+   */
+  async nextSequenceNumber(): Promise<bigint | null> {
+    /* eslint-disable no-await-in-loop */
+    while (this.lock) {
+      await sleep(this.sleepTime);
+    }
+
+    this.lock = true;
+    let nextNumber = BigInt(0);
+    try {
+      if (this.lastUncommintedNumber === null || this.currentNumber === null) {
```

Comment:
So in Python, I got a lint here because I never made an invariant check that these were not null afterward, I ended up adding a boolean for initialized and setting these to 0 (non-nullable) int :facepalm:

Reply:
yes, this is why we have the `!` in https://github.com/aptos-labs/aptos-core/pull/8854/files#diff-ce5e15c21eba1893c7cc44d4dc5a1265a72c0e2490ce84a0f003029f7bd983adR95